Refactor AgencyHomePage to make four possible views explicit

zxt-tzx opened the pull request:
## Problem

To pave the way for easy incorporation of topics management functionality in https://github.com/opengovsg/askgovsg/pull/1350 

Extends https://github.com/opengovsg/askgovsg/pull/1347 and should be merged in after https://github.com/opengovsg/askgovsg/pull/1347 is merged. 

## Solution

Refactor AgencyHomePage to make four possible views explicit

christopherlim98 commented:
Could you change the base branch to #1347  instead of develop so that the diff is easier to read? :)

LoneRifle commented:
> Could you change the base branch to #1347 instead of develop so that the diff is easier to read? :)

ftfy

zxt-tzx committed:
refactor: refactor AgencyHomePage to make four possible views explicit

Main change is to make explicit the components in each the following four views in AgencyHomePage: (1) homePageDefaultView, (2) homePageAllQuestionsView, (3) topicPageDesktopView, (4) topicPageMobileView. Which view is shown depends on urlHasTopicsParamKey and topicQueried, and deviceType === device.desktop.

Other assorted refactoring:
- Use ReactNode instead of ReactChild | ReactChild[] | ReactChildren | ReactChildren[] for conciseness
- Set View All Questions button to only show up by default when user is looking at top Questions -> this is a clearer expression of when the View All Questions button should be shown instead of pagination
- Remove unnecessary use of fetchTopics when agency is undefined; the topics thus fetched are not used anywhere.

LoneRifle reviewed (changes requested):
you need to provide more information in the commit message and/or PR to describe _how_ the refactor was done, as it's not easily inferred from the diff. See https://cbea.ms/git-commit/ for more information.

LoneRifle commented on client/src/components/Questions/Questions.component.tsx:
```diff
@@ -72,7 +72,7 @@ export const Questions = ({
         }
         showViewAllQuestionsButton={
-          showViewAllQuestionsButton ?? questionsDisplayState.value !== 'all'
+          showViewAllQuestionsButton ?? questionsDisplayState.value === 'top'
```
Also, why did you change this?

zxt-tzx replied:
ah this was in response to a bug I found. `showViewAllQuestionsButton` is not a great name (but I was struggling to figure out what a better name is) because what it does is it either shows the "View all questions" button or it shows the pagination. 

After checking out the live AskGov, I noticed the behavior we want is to show the "View all questions" when viewing top questions. 

In the two other display states (i.e. currently viewing all questions OR viewing the questions of a selected topic), we really want to show the pagination. 

Hence this change!

LoneRifle replied:
are you sure about the latter?

zxt-tzx replied:
I... think so? I guess to put it more succinctly, the "View all questions" button should only show up when:
- User is looking at the top questions, so the button brings them to all questions instead
- User is authenticated, in which case they'll be straight-up shown 50 questions, with no need to toggle to see more (I'm assuming the original design intent is 50 questions should be more than enough to see all questions...? I'm guessing this hard-coding might result in errors down the line...?)

In most cases, I don't think it makes a difference because users will typically access topic via the AgencyHomePage, but changing this ensures the following are consistent:

Local without this change:
<img width="946" alt="image" src="https://user-images.githubusercontent.com/67887489/157583594-98be68ce-647b-4f12-853a-7bac13069233.png">

Live AskGov:
<img width="938" alt="image" src="https://user-images.githubusercontent.com/67887489/157584086-1834a9d8-6c51-4338-acb9-e8db9fe1cb45.png">

LoneRifle reviewed (approved):
lgtm, merge when ready
